### PR TITLE
rebase: update kubernetes to v1.21.2

### LINKED ISSUE #1968: update kubernetes dependency to 1.21.0
as kubernetes 1.21.0 is released https://github.com/kubernetes/kubernetes/releases/tag/v1.21.0 we need to update the below items in our codebase

* Update go.mod
* Use kube 1.20 and 1.21 for E2E testing in centos CI
* Update mergify rules for the same.

### PR DESCRIPTION
**Pull request by Rakshith-R**
# Describe what this PR does #
Updated kubernetes packages to latest release.
- resizefs package has been included into k8s.io/mount-util package. updated code to use the same.
- update controller-runtime package. k8s v1.21.2 needs this update to run. Updated code persistentvolume.go Reconcile() and added context.Context accordingly.
- updated controller-runtime to v9.2.0
- e2e: update e2epv.WaitOnPVandPVC() to match new parameters 

fixes: #1968 

~~Dependency: Need controller-runtime v0.9.0 release https://github.com/kubernetes-sigs/controller-runtime/releases 
(Testing with v0.9.0-alpha.1 )~~

TODO: 

- [x] update controller-runtime package to v0.9.0 ~~(not released yet)~~
 
---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

### CODE REVIEW

**Comment by nixpanic** [comment]
@Rakshith-R this is posted as **Draft** and marked with **[DNM]**. Can you add the reason(s) for that in the PR description, and maybe add a TODO list there too?

**Comment by Rakshith-R** [comment]
>Dependency: Need controller-runtime v0.9.0 release https://github.com/kubernetes-sigs/controller-runtime/releases
(Testing with v0.9.0-alpha.1 )

@nixpanic , dependency was mentioned in the description. The issue is the same as here https://github.com/kubernetes-sigs/controller-runtime/issues/1473. I will add a TODO list to it.

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e/k8s-1.21

See https://github.com/ceph/ceph-csi/issues/1963#issuecomment-826585403, maybe it makes a difference.

**Comment by nixpanic** [comment]
> > /retest ci/centos/mini-e2e/k8s-1.21
> 
> See [#1963 (comment)](https://github.com/ceph/ceph-csi/issues/1963#issuecomment-826585403), maybe it makes a difference.

It didn't. Maybe there is a need to update the `libopenstorage/*` package(s)?

**Comment by Madhu-1** [comment]
https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.9.0 is released.

**Comment by Madhu-1** [comment]
update kubernetes to https://github.com/kubernetes/kubernetes/releases/tag/v1.21.1

**Comment by Rakshith-R** [comment]
This is ready for review, DNM label can be removed.
@nixpanic @humblec @Madhu-1

**Comment by Rakshith-R** [comment]
> /retest all

ci/centos/mini-e2e-helm/k8s-1.19 timed out waiting for rbd restored image to be flattened 
+ testing k8s 1.21 e2e

**Comment by Madhu-1** [comment]
> > /retest all
> 
> ci/centos/mini-e2e-helm/k8s-1.19 timed out waiting for rbd restored image to be flattened
> 
> * testing k8s 1.21 e2e

what did the error message, If you see any unusual error message please report it as an issue.

**Comment by Rakshith-R** [comment]
> what did the error message, If you see any unusual error message please report it as an issue.

yes, i would report any unusual error messages. I think in that instance it was just slow ops causing flattening to take long and eventually time out.

**Comment by Rakshith-R** [comment]
- [x] Wait on @humblec to confirm about issues with k8s 1.21.2, to decide to upgrade to which k8s  v1.21.x.

**Comment by humblec** [comment]
>     * [ ]  Wait on @humblec to confirm about issues with k8s 1.21.2, to decide to upgrade to which k8s  v1.21.x.

Sure @Rakshith-R .. will confirm it soon.. On it.

**Comment by humblec** [comment]
> > ```
> > * [ ]  Wait on @humblec to confirm about issues with k8s 1.21.2, to decide to upgrade to which k8s  v1.21.x.
> > ```
> 
> Sure @Rakshith-R .. will confirm it soon.. On it.

OK... Here is the detail about the issue I was mentioning !!


@Rakshith-R 

We shouldnot upgrade to `v1.21.1`. This release has a `regression`  (https://github.com/kubernetes/kubernetes/issues/102480) and there was a notification to **NOT** upgrade to this version. 

Either we have to stay with `1.20.6` or jump directly to `1.21.2` . So, lets upgrade to `1.21.2` and skip `1.21.1`.

Hope it helps

**Comment by Rakshith-R** [comment]
> OK... Here is the detail about the issue I was mentioning !!
> 
> 
> @Rakshith-R 
> 
> We shouldnot upgrade to `v1.21.1`. This release has a `regression`  (https://github.com/kubernetes/kubernetes/issues/102480) and there was a notification to **NOT** upgrade to this version. 
> 
> Either we have to stay with `1.20.6` or jump directly to `1.21.2` . So, lets upgrade to `1.21.2` and skip `1.21.1`.
> 
> Hope it helps

Thanks @humblec 
I think we can wait for 1.21.2 and will be able to upgrade to it in time for v3.4.0 release. 
https://kubernetes.io/releases/patch-releases/#1-21

**Review by Madhu-1 — CHANGES_REQUESTED** [review]
PLease update commit message and PR description/Title

**Comment by Rakshith-R** [comment]
> PLease update commit message and PR description/Title

Updated, thanks 
PTAL @nixpanic @Madhu-1 @humblec

**Review by Madhu-1 — DISMISSED** [review]

File: go.mod [hunk 1]
```diff
@@ -40,28 +39,28 @@ replace (
...
-	k8s.io/kube-controller-manager => k8s.io/kube-controller-manager v0.20.6
-	k8s.io/kube-proxy => k8s.io/kube-proxy v0.20.6
-	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.20.6
-	k8s.io/kubectl => k8s.io/kubectl v0.20.6
-	k8s.io/kubelet => k8s.io/kubelet v0.20.6
-	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.20.6
-	k8s.io/metrics => k8s.io/metrics v0.20.6
-	k8s.io/mount-utils => k8s.io/mount-utils v0.20.6
-	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.20.6
+	k8s.io/api => k8s.io/api v0.21.2
```
Comment (nixpanic) [inline go.mod:42]: can these `replace` entries be removed now? The version in `require` is the same.

Comment (Rakshith-R) [reply go.mod:42]: @nixpanic 
I think we need to still `replace` entries, tried removing all the repeated entries but getting same error as below.
Am I missing/ doing something wrong ? :sweat: 
```
[rakshith@localhost ceph-csi]$ gd
diff --git a/go.mod b/go.mod
index 9c3ac4115..a5017850b 100644
--- a/go.mod
+++ b/go.mod
@@ -39,7 +39,6 @@ replace (
        github.com/hashicorp/vault/api => github.com/hashicorp/vault/api v1.0.5-0.20200902155336-f9d5ce5a171a
        github.com/hashicorp/vault/sdk => github.com/hashicorp/vault/sdk v0.1.14-0.20201116234512-b4d4137dfe8b
        github.com/kubernetes-csi/external-snapshotter/v2 => github.com/kubernetes-csi/external-snapshotter/v2 v2.1.1-0.20200504125226-859696c419ff
-       k8s.io/api => k8s.io/api v0.21.2
        k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.21.2
        k8s.io/apimachinery => k8s.io/apimachinery v0.21.2
        k8s.io/apiserver => k8s.io/apiserver v0.21.2
[rakshith@localhost ceph-csi]$ go mod download
go: k8s.io/kubernetes@v1.21.2 requires
        k8s.io/api@v0.0.0: reading k8s.io/api/go.mod at revision v0.0.0: unknown revision v0.0.0
```

**Comment by Rakshith-R** [comment]
Resolved merge conflicts and updated controller-runtime to v0.9.2(seems to be a quick bug fix). 
PTAL @nixpanic @Madhu-1

File: go.mod (hunk 1 above)
Comment (Madhu-1) [reply go.mod:42]: Replace is required to satisfy internal dependency what I have seen earlier was without this one, the latest dependency will get pulled, or the build will fail.

**Review by Madhu-1 — DISMISSED** [review]

**Comment by Rakshith-R** [comment]
hey @nixpanic @humblec , PTAL 
and call mergifyio rebase :slightly_smiling_face: 
e2e failed because of 
`Jun 25 06:41:29.302: INFO: Error getting pvc in namespace: '': etcdserver: request timed out
Jun 25 06:41:29.302: INFO: failed to create PVC (rbd-6940): failed to get pvc: etcdserver: request timed out`

File: go.mod (hunk 1 above)
Comment (humblec) [reply go.mod:42]: yeah, replace is required,

**Comment by humblec** [comment]
Sure @Rakshith-R reviewing this PR

**Review by humblec — DISMISSED** [review]

**Comment by humblec** [comment]
@Mergifyio rebase

**Comment by Rakshith-R** [comment]
Addressed `lll` golang ci linter issue and rebased on top of devel.  :sweat_smile: 
PTAL @nixpanic @Madhu-1 @humblec

**Review by Madhu-1 — DISMISSED** [review]

**Comment by Madhu-1** [comment]
@Rakshith-R CI is failing make sure you run the tests locally and it's passing.

**Comment by Rakshith-R** [comment]
> Addressed `lll` golang ci linter issue and rebased on top of devel. 
> PTAL @nixpanic @Madhu-1 @humblec

Resolved the ones that got away earlier too. PTAL.

**Comment by Madhu-1** [comment]
> > Addressed `lll` golang ci linter issue and rebased on top of devel.
> > PTAL @nixpanic @Madhu-1 @humblec
> 
> Resolved the ones that got away earlier too. PTAL.

Thanks 😊

**Review by Madhu-1 — APPROVED** [review]

**Review by nixpanic — APPROVED** [review]

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e-helm/k8s-1.21

Failed to deploy Ceph cluster ([logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.21/detail/mini-e2e-helm_k8s-1.21/48/pipeline))

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e/k8s-1.21

Failed really early in the testing, on initial namespace creation ([logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.22/detail/mini-e2e-helm_k8s-1.22/3/pipeline))

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e/k8s-1.19

#1969 happened again ([logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.19/detail/mini-e2e_k8s-1.19/1826/pipeline))

**Comment by Rakshith-R** [comment]
@mergifyio rebase

**Comment by Rakshith-R** [comment]
> @Mergifyio rebase

mergyfio was complaining about changes in its configuration and needing manual action.

 `The configuration has been changed, queue and merge actions are ignored. :warning`
https://github.com/ceph/ceph-csi/pull/1974/checks?check_run_id=2951833427

**Comment by Rakshith-R** [comment]
@Mergifyio rebase

Mini e2e 1.19 https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.19/detail/mini-e2e_k8s-1.19/1834/pipeline

Mini e2e 1.21
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e_k8s-1.21/55/display/redirect

Both failed due to `etc server request timedout`

Using rebase since another pr got merged in the meantime.

**Pull request merged**